### DISCUSSION TITLE
End trial after response but display stimulus for stimulus_duration

### DISCUSSION THREAD

**Discussion by JAQuent** [jspsych/jsPsych#1524] [Q&A] [answered]
Hey, I like to use the jspsych-html-keyboard-response plugin and create a trial that gives the participant 10 secs to respond. The stimulus_durations shoud be 800 msec. 

If a response is given before the 800 msec are over, then the stimulus should stay on until that is the case and then end. If a response is given after 800 msec, then I want the trial to end.

I played around with `stimulus_duration`, `trial_duration` and `response_ends_trial` but if I set the latter to true than a response also ends stimulus presentation. The description on those arguments (https://www.jspsych.org/plugins/jspsych-html-keyboard-response/) are ambiguous on this.

Is that even possible?

**Comment by JAQuent** [comment, marked as answer]
I've made custom edits to `end_trial() ` function and I think it works. 

The main change inside that function is:
```
      // if responses_ends_trial_after_stimulus_duration is true wait until the stumulus_duration is over
      if(trial.responses_ends_trial_after_stimulus_duration){
        var remaining_time = trial.stimulus_duration - response.rt;
         jsPsych.pluginAPI.setTimeout(function() {
          // clear the display
          display_element.innerHTML = '';

          // move on to the next trial
          jsPsych.finishTrial(trial_data);
        }, remaining_time);
      } else {
        // clear the display
        display_element.innerHTML = '';

        // move on to the next trial
        jsPsych.finishTrial(trial_data);
      }
```

I am wondering whether this is worth adding as a feature. In a lot of memory experiments you want the stimulus duration to be fixed but give the participants more time to respond as this ensures equal encoding time for each stimulus but at the same time end the trial as soon as a response is given.

**Reply by becky-gilbert** [reply]
Hi @JAQuent, sorry for the delay - we're busy trying to finish up a release right now. Glad you were able to figure this out!

You're right that there's currently no parameterized of doing this in a single trial. Without modifying the plugin, the only other option is to use two separate trials - one to show the stimulus for a fixed duration, with `response_ends_trial: false`, and a blank trial that just runs if no response was made in the trial, with `response_ends_trial: true`. But that's not an ideal solution...

I agree that your new parameter might be useful to others. Please feel free to create a new issue with this suggestion and/or submit a pull request! 👍 

In terms of implementation, it looks like your additional `setTimeout` timer for `remaining_time` will run even if the response occurs _after_ the `stimulus_duration`, which may not be ideal. So you could add a check to see if a response was made and `remaining_time` is positive before setting up the timer. Another suggestion is to move the `clearAllTimeouts` function into the bits of code that actually end the trial in the `end_trial` function. So here's my version:
```js
var end_trial = function() {

  // kill keyboard listeners
  if (typeof keyboardListener !== 'undefined') {
    jsPsych.pluginAPI.cancelKeyboardResponse(keyboardListener);
  }

  // gather the data to store for the trial
  var trial_data = {
    rt: response.rt,
    stimulus: trial.stimulus,
    response: response.key
  };

  // if response_ends_trial_after_stimulus_duration is true and a response was made before stimulus_duration,
  // then wait until the stimulus_duration is over before ending the trial
  if (trial.response_ends_trial_after_stimulus_duration && response.rt !== null && response.rt < trial.stimulus_duration) {
    var remaining_time = trial.stimulus_duration - response.rt;
    jsPsych.pluginAPI.setTimeout(function() {
      // kill any remaining setTimeout handlers
      jsPsych.pluginAPI.clearAllTimeouts();
      // clear the display
      display_element.innerHTML = '';
      // move on to the next trial
      jsPsych.finishTrial(trial_data);
    }, remaining_time);
  } else {
    // kill any remaining setTimeout handlers
    jsPsych.pluginAPI.clearAllTimeouts();
    // clear the display
    display_element.innerHTML = '';
    // move on to the next trial
    jsPsych.finishTrial(trial_data);
  }
};
```
And another implementation issue: it might be worth moving this stuff out of the `end_trial` function, so that the `end_trial` function just ends the trial, and to avoid the repeated code inside this function. So a different approach would be to keep the `end_trial` function as it was before, and use the `setTimeout` for the `stimulus_duration` to see whether a response has been made and `response_ends_trial_after_stimulus_duration` is `true`, in which case the trial should end:
```js
// hide stimulus if stimulus_duration is set
if (trial.stimulus_duration !== null) {
  jsPsych.pluginAPI.setTimeout(function() {
    // hide the stimulus
    display_element.querySelector('#jspsych-html-keyboard-response-stimulus').style.visibility = 'hidden';
    // if response_ends_trial_after_stimulus_duration is true and a response has been made, then end the trial now
    if (trial.response_ends_trial_after_stimulus_duration && response.rt !== null) {
      end_trial();
    }
  }, trial.stimulus_duration);
}
```
And with this approach, you'd also need to change the `after_response` function to make sure that the trial ends after a response if the response is made _after_ `stimulus_duration`:
```js
var after_response = function(info) {

  // after a valid response, the stimulus will have the CSS class 'responded'
  // which can be used to provide visual feedback that a response was recorded
 display_element.querySelector('#jspsych-html-keyboard-response-stimulus').className += ' responded';

  // only record the first response
  if (response.key == null) {
    response = info;
  }
  // end trial immediately if response_ends_trial is true
  if (trial.response_ends_trial) {
    end_trial();
  }
  // end trial immediately if response_ends_trial_after_stimulus_duration and the response was made after stimulus_duration
  if (trial.response_ends_trial_after_stimulus_duration && response.rt > trial.stimulus_duration) {
    end_trial();
  }
};
```
Anyway, feel free to ignore these suggestions - just wanted to share in case you're interested 😃

**Reply by JAQuent** [reply]
Those suggestion are good! I will probably submit a pull request next week. I've never done this so I'll need to figure out how it works first.

**Reply by becky-gilbert** [reply]
Yay, great! There's some [information about contributing](https://www.jspsych.org/about/contributing/#fork-the-library-and-modify-the-code) on the website, in case you missed it. Let us know if you have questions.

We'd appreciate it if your pull request could include any updates to the [plugin docs](https://github.com/jspsych/jsPsych/blob/master/docs/plugins/jspsych-html-keyboard-response.md) to reflect your changes to the plugin code. And don't forget to add yourself to the [contributors list](https://github.com/jspsych/jsPsych/blob/master/contributors.md)!
